### PR TITLE
SEGA Saturn / Super-H architecture

### PR DESCRIPTION
**Pull request by Fothsid**
Mainly adds the SEGA Saturn / Super-H 2 architecture & instruction set support. (`.saturn` directive)
The code was based on the MIPS parser.

Minor changes include:
- Added a fallback to the ELF relocator for PSX architecture, in case if the PSX relocator doesn't succeed.
- Changed ELF file handling code to support RELA sections (REL, but with an addend)

### CODE REVIEW

**Comment by Fothsid** [comment]
I see the compilation issues with MSVC, will try to patch them up soon.

**Comment by sp1187** [comment]
Some additional test cases, preferably at least one of every single opcode would be nice to ensure its encoding is correct (no typo regressions later etc.).

**Comment by Fothsid** [comment]
Made a test for every SH1/SH2 opcode, turned out there were some bugs, which I fixed. Expected output was taken from GCC's SH assembler.

File: Commands/CDirectiveFile.cpp [hunk 1]
```diff
@@ -512,15 +512,18 @@ void CDirectiveHeaderSize::writeTempData(TempData& tempData) const
 DirectiveObjImport::DirectiveObjImport(const fs::path& inputName)
 {
 	ctor = nullptr;
-	if (rel.init(inputName))
+	success = rel.init(inputName);
+	if (success)
 	{
+		success = true;
```
Comment (Kingcom) [inline Commands/CDirectiveFile.cpp:518]: `success` is already true at this point.

**Review by Kingcom — CHANGES_REQUESTED** [review]

File: Archs/SuperH/CShInstruction.cpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+	this->immediateData = immediate;
+	this->registerData = registers;
+}
+
+CShInstruction::~CShInstruction()
+{
+
+}
+
+int getImmediateBits(ShImmediateType type)
```
[the remark this answers is not in the document]
Comment (Fothsid) [reply Archs/SuperH/CShInstruction.cpp]: Done.

File: Archs/SuperH/CShInstruction.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+//       Delayed opcodes are currently marked with the
+//       SH_DELAYED flag.
+//
+
+bool CShInstruction::Validate(const ValidateState &state)
+{
+	RamPos = g_fileManager->getVirtualAddress();
+	if (RamPos % 2)
+	{
+		Logger::queueError(Logger::Error, "opcode not aligned to word boundary");
```
[the remark this answers is not in the document]
Comment (Fothsid) [reply Archs/SuperH/CShInstruction.cpp:48]: Yes. On Super-H, a word is a 16-bit value. Instructions like `mov.w` store/load 16-bit values, whereas `mov.l` does 32-bit values.

File: Archs/SuperH/ShOpcodes.cpp [hunk 4]
```diff
@@ -0,0 +1,204 @@
+#include "Archs/SuperH/ShOpcodes.h"
+
+const tShOpcode shOpcodes[] = {
+	/*
+		I'm not entirely sure whether having a hash before an immediate
+		byte value in an instruction should be a requirement. As far as examples go,
+		it seems like Ghidra uses a hash for the immediate mov instruction, but doesn't
+		use it for the add instruction.
+
+		I think I'll have all immediate byte values require a hash before it, just for
```
[the remark this answers is not in the document]
Comment (Fothsid) [reply Archs/SuperH/ShOpcodes.cpp]: Added aliases for the instructions with the immediate 8-bit values.

**Comment by sp1187** [comment]
Another thing I thought of: Since the Sega 32X uses the same SH2 CPU, do you think it would make sense to add a `.32x` alias directive that maps to the same action as `.saturn`?

**Comment by Fothsid** [comment]
That does sound interesting. I completely forgot that 32x is a thing, haha. I'll probably add it in a bit.

**Comment by Kingcom** [comment]
Is there any odd memory mapping on the 32X, like on SNES? If there isn't full support for the system or it needs workarounds, it should be mentioned in the Readme. Can you also add the directive to it?

**Pull request merged**